GoogleForCreators/web-stories-wp
# Page Attachment as linked content

## Summary

Adds the Page Attachment features as linked content.


## Relevant Technical Choices



## To-do

- [x] Needs UX for the general behavior in case a link ends up in an invalid position.

- [x] UX adjustments for the editor

- [x] Add "default" suffix to Learn more

- [x] Behavior when there is a linked element

- [x] Refactor code a bit

- [x] Disable links for output that are in the Page Attachment area.

- [x] Tests (Need update post-UX change)



## User-facing changes

There is a new Page Attachment panel which the user can add while Page/background is selected.
The moment when the user adds a URL, the Page Attachment gets added.


## Testing Instructions
I
1. Open a Page
2. Add a URL to the Page Attachment
3. Verify that if you add incorrect URL format (e.g. with a space inside), a warning is displayed

II
1. Open a Page
2. Add a URL to the Page Attachment
3. Verify that "Learn more" appears in the bottom of the page
4. Change the "Learn more" to something else.
5. Verify that it also changed on the Page.
6. Verify that removing the CTA Text will restore the default "Learn more"
7. Verify that the Page Attachment is also displayed in the Front-end and with the correct CTA Text.


III
1. Open a Page
2. Add a URL to the Page Attachment
3. Add a Shape, add a link to the Shape
4. Try dragging/resizing and rotating the link so that part of the element is near the bottom -- where the "Learn more" is displayed.
5. Verify that after finishing with the transformation, a dashed line appears with a tooltip on the left saying that links can't be below the dashed line with Page Attachment.

IV
1. Open a Page
2. Add a URL to the Page Attachment
3. Add 3 elements, add link to one of these elements
4. Select all 3 elements
5. Drag/resize/rotate the elements so that the linked element is positioned near the "Learn more" text.
6. Verify again the same dashed line with the tooltip.

V
1. Add a shape with link
2. Move the shape to the bottom area of the page
3. Now click on the Page and click on the Page Attachment input without typing anything.
4. Verify that the dashed line is displayed again and a warning message appears explaining that links of the elements below the dashed line won't be clickable in case of adding Page Attachment.
5. Add a Page Attachment URL, save the story and verify that the link is not clickable indeed.

VI
1. Add a Page Attachment
2. Create a shape and move it to the Page Attachment area.
3. Focus on the link field of the Shape element.
4. Verify that a warning appears when focusing on the link input and the dashed line is displayed.
5. Verify that it's not possible to fill in the link field.

---



Fixes #254
